File: 2024 1월 3주차/백준 1012.java
import java.util.Scanner;

public class Main {
    static int[] dx = {1, -1, 0, 0};
    static int[] dy = {0, 0, 1, -1};
    static int n, m;
    static int[][] field;
    static boolean[][] visited;

    public static void dfs(int x, int y) {
        visited[x][y] = true;
        for (int i = 0; i < 4; i++) {
            int nx = x + dx[i];
            int ny = y + dy[i];
            if (0 <= nx && nx < n && 0 <= ny && ny < m && field[nx][ny] == 1 && !visited[nx][ny]) {
                dfs(nx, ny);
            }
        }
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int t = scanner.nextInt();
        for (int tc = 0; tc < t; tc++) {
            m = scanner.nextInt();
            n = scanner.nextInt();
            int k = scanner.nextInt();
            field = new int[n][m];
            visited = new boolean[n][m];

            for (int i = 0; i < k; i++) {
                int y = scanner.nextInt();
                int x = scanner.nextInt();
                field[x][y] = 1;
            }

            int count = 0;
            for (int i = 0; i < n; i++) {
                for (int j = 0; j < m; j++) {
                    if (field[i][j] == 1 && !visited[i][j]) {
                        dfs(i, j);
                        count++;
                    }
                }
            }
            System.out.println(count);
        }
        scanner.close();
    }
}

File: 2024 1월 3주차/백준 1012.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

int dx[4] = {1, -1, 0, 0};
int dy[4] = {0, 0, 1, -1};
int n, m;

void dfs(int x, int y, int** field, bool** visited) {
    visited[x][y] = true;
    for (int i = 0; i < 4; i++) {
        int nx = x + dx[i];
        int ny = y + dy[i];
        if (0 <= nx && nx < n && 0 <= ny && ny < m && field[nx][ny] == 1 && !visited[nx][ny]) {
            dfs(nx, ny, field, visited);
        }
    }
}

int main() {
    int t;
    scanf("%d", &t);
    while (t--) {
        int k;
        scanf("%d %d %d", &m, &n, &k);

        int** field = (int**)malloc(n * sizeof(int*));
        bool** visited = (bool**)malloc(n * sizeof(bool*));
        for (int i = 0; i < n; i++) {
            field[i] = (int*)malloc(m * sizeof(int));
            visited[i] = (bool*)malloc(m * sizeof(bool));
            for (int j = 0; j < m; j++) {
                field[i][j] = 0;
                visited[i][j] = false;
            }
        }

        for (int i = 0; i < k; i++) {
            int x, y;
            scanf("%d %d", &y, &x);
            field[x][y] = 1;
        }

        int count = 0;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (field[i][j] == 1 && !visited[i][j]) {
                    dfs(i, j, field, visited);
                    count++;
                }
            }
        }
        printf("%d\n", count);

        for (int i = 0; i < n; i++) {
            free(field[i]);
            free(visited[i]);
        }
        free(field);
        free(visited);
    }
    return 0;
}